GetStream/stream-chat-android
# [3963] Side effects for DistinctCalls run multiple times

### 🎯 Goal

Make side effects for DistinctCalls run once.

### 🛠 Implementation details

A new operator named `share` has been introduced to reuse the same chain of the calls as long as the real request keeps running.

### 🧪 Testing

- Run sample app
- Open any channel 
- Make sure there is a single `doOnStart` and `doOnResult` presented in the logs even though the `queryChannel` is invoked multiple times

```
2022-08-18 19:29:49.657 5359-5359/io.getstream.chat.ui.sample.debug D/Chat:Client: (main:2) [queryChannel] cid: messaging:sample-app-channel-109
2022-08-18 19:29:49.661 5359-5470/io.getstream.chat.ui.sample.debug V/Chat:Client: (DefaultDispatcher-worker-13:119) [queryChannel] #doOnStart; plugin: io.getstream.chat.android.offline.plugin.internal.OfflinePlugin
2022-08-18 19:29:49.668 5359-5359/io.getstream.chat.ui.sample.debug D/Chat:Client: (main:2) [queryChannel] cid: messaging:sample-app-channel-109
2022-08-18 19:29:49.678 5359-5359/io.getstream.chat.ui.sample.debug D/Chat:Client: (main:2) [queryChannel] cid: messaging:sample-app-channel-109
2022-08-18 19:29:49.685 5359-5359/io.getstream.chat.ui.sample.debug D/Chat:Client: (main:2) [queryChannel] cid: messaging:sample-app-channel-109
2022-08-18 19:29:49.718 5359-5359/io.getstream.chat.ui.sample.debug D/Chat:Client: (main:2) [queryChannel] cid: messaging:sample-app-channel-109
2022-08-18 19:29:50.488 5359-5587/io.getstream.chat.ui.sample.debug V/Chat:Client: (DefaultDispatcher-worker-76:215) [queryChannel] #doOnResult; plugin: io.getstream.chat.android.offline.plugin.internal.OfflinePlugin

```


### ☑️Contributor Checklist

#### General

#### Code & documentation

### ☑️Reviewer Checklist

### 🎉 GIF

_Please provide a suitable gif that describes your work on this pull request_
